### ISSUE TITLE
Adding custom keystrokes through Lua

### ISSUE THREAD

**Issue by KenneyNL** [minetest/minetest#2731]
To add advanced functionality to mods adding custom keystrokes would be great. Bundling that together with pointing at nodes it could be used to give nodes more than 1 action (knocking on a door, opening/closing a door and locking/unlocking a door).

``` lua
minetest.register_key({
   key = 'C',
   function = do_c()
})
```

Previous discussion:
https://forum.minetest.net/viewtopic.php?f=5&t=11018

**Comment by C1ffisme** [comment]
How many feature requests are you going to make today?

No, wait, I'm being a hypocrite. I think I've made more in a small space of time.

**Comment by KenneyNL** [comment]
Hah well I've posted these before on the forums and some I just encountered while modding. This is all...for today.

**Comment by C1ffisme** [comment]
Ok.

**Comment by TeTpaAka** [comment]
#2445 is related.

**Comment by est31** [comment]
This needs clientside scripting. And it perhaps shouldn't be called `function`, `callback` is better. Also, it would be faster to not pass a table (which is a hashmap), but a list of arguments.

**Comment by 4aiman** [comment]
Has been discussed countless number of times...

Just add more auxiliary keys. 
6 instead of 2 would be *more* than enough. 

But that would be a protocol bump.
And that's the main reason devs probably won't want this till 0.5...

**Comment by Calinou** [comment]
I agree, having 4-6 special keys is a good solution for now; this has the upside of letting users change their keys and not breaking some keyboard layouts like AZERTY.

**Comment by est31** [comment]
Adding `get_player_control` has been a mistake, which is broken by design.

**Comment by 4aiman** [comment]
@est31 
So... what is *not* a mistake?
Since ```get_player_control``` is very useful, do you have any thoughts on how to replace that?

**Comment by est31** [comment]
The problem is that these controls get sent to the server, and are subject to the lag. This is ok for a shell, but not at all ok for a game, where you for example want to control a vessel.

**Comment by 4aiman** [comment]
@est31 
Is that so?
I mean, it's ```getter```, *not*  ```setter```.
Setting ```player:get_player_controls().sneak = true``` has absolutely no effect on server.
The idea of adding
```
minetest.register_key({
   key = 'C',
   function = do_c()
})
```
*will* be broken by design, but a getter for currently pressed keys is not.
Of do you have the fact of sending current keys *from* sever?

Anyway, w/o client-side modding there won't be any "normal" way to control a vessel.
Or will?
I mean, if the current design *is* broken, then it needs to get fixed asap - it's not a subgame's weirdness but lag that stirs players away from a server.

**Comment by est31** [comment]
Until client side modding arives, I won't remove that function. But it shouldn't be extended.

**Comment by 4aiman** [comment]
@est31 
So, how about adding more keys that will toggle some setting?
Like binding, say, the "K" and "L" keys to control some existing variable from the config (gui_scale for example)? 

This won't be a client-side modding and won't add any server lag, but will enable users to control mods that check those variables in real-time.

This won't allow anything non-client to be sent to a server - only local setting will be affected.

**Comment by est31** [comment]
related to #560

**Comment by est31** [comment]
@4aiman the game should still be playable on mobile devices, for example. If we allow to map *every* key, we not just lose the ability to assign new keys to tasks in the future (because that would break mods), players using won't be abled to use those features. Look at the [hungry games mod](https://forum.minetest.net/viewtopic.php?f=3&t=12430) as an example, and thats only about the use key.

**Comment by 4aiman** [comment]
@est31 That's really a problem, though not in the case you provided a link for.

Sprinting can be done MC-way and it will work on any device then (double-tapping "up" works for me ;)

As for mobile devices (e.g. android powered) and the problem of auxiliary keys in general - I can't see how additional keys can't be added to the ```touchscreengui``` module.

Moreover, it would be just *great* to be able to plug-in a keyboard and a mouse and be able to control the game via those manipulators.

*Most* of cheap android devices can't handle minetest in ```singleplayer``` well enough.
*Many* of cheap android devices can't handle minetest in ```multiplayer``` well enough.

The default minetest_game is rather "ok" in multiplayer for a device with ARMv7 1200MHz CPU and *at least* 512Mb of RAM. 
For a singleplayer one will need 2x1GHz + *at least* 512Mb of RAM device (because of mapgen and mods like "fire").

And if one create a separate subgame which adds different features like mobs, potions, tech, achievements, use mgv7 - one will need even more CPU power and RAM. 

Right now only powerful devices can handle any minetest game like they should.

-----------------

**TL;DR**

1. I acknowledge the problem exists, but there's such thing as the ```minimal system requirements```. 

2. Every mobile device should either be supported or not - MT for semi-supported devices should not exist.

**Comment by est31** [comment]
Related: #1693 #1737

**Comment by paramat** [comment]
Folded #1693 into this.

**Comment by ObaniGemini** [comment]
I think we can also talk about https://github.com/minetest/minetest/pull/3856

**Comment by est31** [comment]
@ObaniGemini that's something different, its about shortcuts set by the client.

**Comment by 4aiman** [comment]
I think this should be closed if there won't be 2 or more approvals. @est31 gave us his (sorry if I "misguessed" the pronoun) opinion and it hasn't changed (as far as I can judge). 

There are plenty of requests like this: old, discouraged and almost "won't add/fix", repeated multiple times since early 0.4-delta... It's better to close all of those and re-open if needed later.

**Comment by PureTryOut** [comment]
Yes please! Maybe the mod can say which specific buttons it wants to listen too, and the client then checks if one of those is pressed. If so, it sends this back to the server running the mod.
I have no clue how this would be done though, is it impossible with how the game works right now?

**Comment by Avamander** [comment]
One way to solve this would be to add the functionality, but tell the mobile users of those particular servers that their client might not support certain special keys. This would definitely not be ideal for mobile users, but PC users don't have to suffer.

**Comment by purveyor-of-points** [comment]
Sorry for the somewhat old revive, but I'd really like to see some degree of implementation still- I'd settle for more special keys as a temporary fix, but it'd be great to have more keys to work with as a mod creator.

For example, being able to have a "Sprint" key, a "Use" key (for knocking on doors, separating it from right click, etc), and a "Grab" key (For grabbing ledges and pulling the character up) would be excellent. There's so many possible controls one could add for some great mods (Combat, movement, etc)- I was honestly surprised to see that it's not already a thing.

**Comment by tobyplowy** [comment]
@KenneyNL +1

**Comment by paramat** [comment]
Any more comments?
Is anyone interested in coding this sometime? if not closing in 1 month.

**Comment by Avamander** [comment]
@paramat Please do not close issues because noone is going to implement it right now. It has to stay open to have a proper overview of needed feature requests. Closing for the sake of closing is useless and won't benefit you the slightest.

**Comment by basicer** [comment]
We have plans to implement this into Client Side Scripting.

Currently there's a rough version of this working internally, but it's not ready for upstream yet.

**Comment by paramat** [comment]
Yeah i'm not doing the 'close in 1 month' thing now.

**Comment by paramat** [comment]
Any core dev interest?

**Comment by rubenwardy** [comment]
Yes

**Comment by runsy** [comment]
Important for me :-)

**Comment by Calinou** [comment]
@runsy Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by runsy** [comment]
> 
> 
> @runsy Please don't bump issues without contributing significant new information. Use the 👍 reaction button on the first post instead.

SORRY :-(

**Comment by rubenwardy** [comment]
Closing in favour of #12488 because it has more information
